### ISSUE TITLE
Bazel Developer Documentation is Drafted

### ISSUE THREAD

**Issue by nstng** [magma/magma#11917]
# How to Bazel@Magma

:warning: This is an initial draft for a Bazel documentation aiming for developers at Magma. This documentation should be moved to a more appropriate place, e.g., github wiki or even docusaurus when sufficiently matured. :warning:

This is a documentation for Bazel at Magma aiming at developers.

For a more in depth documentation see the [official Bazel reference](https://bazel.build/reference).

## Introduction

Bazel is a build system designed for mono-repos. Among others, features are multi language support, reproducible builds by design, extensive build caching and a strict dependency management. Build definitions in Bazel are written in [Starlark](https://github.com/bazelbuild/starlark). 

You can build `mobilityd` by
`~/magma$ bazel build //lte/gateway/python/magma/mobilityd:mobilityd`
	
Now you can find an executable
`~/magma/bazel-bin/lte/gateway/python/magma/mobilityd/mobilityd`
beside needed runtime libraries (internal and external).

In general Bazel supports various verbs (`build`, `run`, `test`, ...) that are used in most cases as follows:
`bazel <verb> <target>`
A target can be for example a library, a binary or a test. A target is defined by a rule, depending on it's purpose (e.g., a C library or a Python binary). In most cases these are binaries, libraries or tests for different languages. Most rules are loaded from external sources located at github.

## Contact Channels

If you create a PR with Bazel changes then Bazel code owners get notified. This will be changed back to the respective component owners when Bazel is sufficiently matured at Magma. For general support there is the Slack channel **#bazel**.

## Structure

Small extract of the Bazel file structure:

```
.
├── bazel
│   ├── bazelrcs
│   │   ├── devcontainer.bazelrc
│   │   └── ...
│   ├── external
│   │   ├── BUILD.bazel
│   │   ├── sentry_native.BUILD
│   │   └── ...
│   ├── BUILD.bazel
│   ├── cpp_repositories.bzl
│   ├── go_repositories.bzl
│   ├── python_repositories.bzl
│   ├── python_swagger.bzl
│   └── ...
├── lte
│   ├── gateway
│   │   └── python
│   │       ├── magma
│   │       │   ├── mobilityd
│   │       │   │   ├── tests
│   │       │   │   │   └── BUILD.bazel
│   │       │   │   └── BUILD.bazel
│   │       │   └── policydb
│   │       │       ├── servicers
│   │       │       │   └── BUILD.bazel
│   │       │       ├── tests
│   │       │       │   └── BUILD.bazel
│   │       │       └── BUILD.bazel
│   │       └── BUILD.bazel
│   ├── protos
│   │   ├── oai
│   │   │   └── BUILD.bazel
│   │   └── BUILD.bazel
│   └── swagger
│       └── BUILD.bazel
├── .bazelrc
├── BUILD.bazel
└── WORKSPACE.bazel
```

* `lte/gateway/python/magma/mobilityd/BUILD.bazel`
  * A Bazel BUILD file. Here usually sources are defined in binaries and libraries. A BUILD file defines a package.
* `WORKSPACE.bazel`
  * Defines a Bazel project. Configuration of rules to be used in the project.
* `.bazelrc`
  * General configuration of flags for verbs.
* `bazel/`
  * Centralized folder for custom Bazel definitions and configuration of external sources.
* `bazel/bazelrcs/name.bazelrc`
  * Configuration of flags for verbs in different environments.
* `bazel/name.bzl`
  * Custom Bazel code - these are custom rules, configuration of the tool chain and external dependencies.
  * In `cpp`, `go` and `python_repositories.bzl` are external repositories specified. These can be for example local repositories (folder in the environment) or git repositories.
* `bazel/external/name.BUILD`
  * BUILD files for external repositories that are not Bazelified.

## Build Targets

A target definition strongly depends on the rule that is used. Targets are specified in [Starlark](https://github.com/bazelbuild/starlark) syntax. Example
```
cc_library(
	name = "my_library",
	srcs = ["my_code.c"],
	hdrs = ["my_header.h"],
	deps = ["//other/path:other_library"],
)
```
A C-library with source file, header and a build dependency to another library. The attribute `name` is the name of the target and is used for building the target directly (`bazel build //lib/path:my_library`) or for referencing as a dependency in other targets.

## Building

Targets are build in Bazel via `bazel build [<target>...]`, e.g.,

`bazel build //lte/gateway/python/magma/mobilityd:mobilityd`

Building happens in a sandbox, i.e., various host properties are not present at build time. This includes environment variables and `stdin`. The build artifact can be found in `$MAGMA_ROOT/bazel-bin`, e.g., `$MAGMA_ROOT/bazel-bin/lte/gateway/python/magma/mobilityd/mobilityd`.

### Notes for Targets in the Command Line

* The leading `//` are redundant, but make it more clear that a target is referenced as the prefix is needed in build files when a dependency is referenced.
* `bazel build //lte/gateway/...` - builds all targets in `lte/gateway`
* `bazel build ...` - builds all targets
* `bazel build //lte/gateway/python/magma/mobilityd` - builds the target that has the same name as the last folder, i.e., here `:mobilityd`.

## Testing

Tests are also defined as targets and can be executed via `bazel test`. If the `...` expansion is used Bazel will only execute test targets, i.e., with `bazel test ...` all Bazel tests are executed (C, Go and Python). `bazel test //lte/...` executes all `lte` tests and `bazel test //lte/gateway/python/...` all `lte` Python tests.

With `bazel coverage <target>` test coverage can be generated. Coverage data can be found in `$MAGMA_ROOT/bazel-out/_coverage/_coverage_report.dat` and further processed with, e.g., `genhtml -o genhtml bazel-out/_coverage/_coverage_report.dat` (`genhtml` can be installed via `apt install lcov`).

## Adding Dependencies

A target that can work with dependencies (all libraries, binaries and tests) usually has a property `deps` - a list of dependencies. Here dependencies can be specified:
```
python_library(
  name = "my_lib",
  srcs = ["my_lib.py"],
  deps = [
  	":my_package_dependency",
  	"//orc8r/magma/common:my_util",
  	"@aioh2_repo//:aioh2",
  	requirement("grpcio"),
  ],
)
```

### Local Dependencies

Dependencies that are specified in the same package (same `BUILD.bazel`) can be added by the name of the target prefixed by `:`. Dependencies on targets in other packages are additionally prefixed by `//` and the relative path to the respective package.

### External Dependencies

External dependencies can be added to `$MAGMA_ROOT/bazel/<language>_repositories.bzl`. The dependency is prefixed by `@` and the name of the dependency and followed by `:` and the name of the target. The target is either specified in the external source (if it is a Bazelified dependency) or in the respective build file in `$MAGMA_ROOT/bazel/external` (if it is a dependency that is not Bazelified).

### Pip Dependencies (Python)

External dependencies in Python can be added in most cases as pip dependencies. For this the dependency needs to be added in `$MAGMA_ROOT/bazel/external/requirements.in`. If a new dependency is added, the `requirements.txt` needs to be rebuild by `$MAGMA_ROOT/bazel/external $ pip-compile --allow-unsafe --generate-hashes --output-file=requirements.txt requirements.in`. The dependency can now be added to a target by using, e.g., `requirement("grpcio")`.

## Environments

Environments where Bazel can be used are getting currently extended and polished.

Bazel is supported on:
* The Dev-Container (see `$MAGMA_ROOT/.devcontainer` for configuration).
  * Note that on the devcontainer various packages (esp. Python dependencies) are pre-installed, i.e., currently here builds can work where other environments might need specified dependencies.
* For codespaces holds the same as for the Dev-Container.
* The Magma-VM.
* The [Bazel base container](https://github.com/magma/magma/pkgs/container/magma%2Fbazel-base) (see `$MAGMA_ROOT/experimental/bazel-base`). This is a minimalistic container where only needed dependencies are installed. This is, this environment can be used to test if all needed dependencies are specified.

Bazel is _not_ supported on:
* A local setup might be possible but a container or VM based environment is recommended so that dependencies installed locally do not interfere with Bazel builds.
* The Test-VM currently does not support Bazel. This is because currently Python 3.8 is needed as a dependency on the used environment (the Test-VM uses 3.5).

**Comment by themarwhal** [comment]
Thanks for starting this! A few notes:

1. I think we should have a section on different build environments we support. (Magma dev VM, GitHub Codespaces/devcontainer, bazel-base containeg)
2. In Structure section, we should mention cpp_repositories.bzl/python_repositories.bzl/requirements as that is where dependencies are managed.
3. I think it would be good to have a section on dependency management for each language. How to add a new pip dependency for a Python target? C/C++? Golang?

**Comment by nstng** [comment]
> A few notes:

@themarwhal applied your notes - if you have more comments, let me know.

**Comment by nstng** [comment]
closed in favor of #14653
